File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/TestSkillPredicate.java
package rustleund.fightingfantasy.framework.closures.impl;

import org.w3c.dom.Element;

import rustleund.fightingfantasy.framework.base.GameState;
import rustleund.fightingfantasy.framework.util.DiceRoller;

import com.google.common.base.Predicate;

public class TestSkillPredicate implements Predicate<GameState> {

	private int diceRollAdjustment;

	public TestSkillPredicate(Element element) {
		if (element.hasAttribute("diceRollAdjustment")) {
			this.diceRollAdjustment = Integer.parseInt(element.getAttribute("diceRollAdjustment"));
		}
	}

	@Override
	public boolean apply(GameState gameState) {
		return (DiceRoller.rollDice(2) + diceRollAdjustment) <= gameState.getPlayerState().getSkill().getCurrentValue();
	}

}

File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/TestSkillPredicate.kt
package rustleund.fightingfantasy.framework.closures.impl

import org.w3c.dom.Element
import rustleund.fightingfantasy.framework.util.DiceRoller.rollDice
import rustleund.fightingfantasy.framework.base.GameState
import rustleund.fightingfantasy.framework.base.AbstractEntityState
import rustleund.fightingfantasy.framework.base.optionalAttribute
import java.util.function.Function
import java.util.function.Predicate

class TestSkillPredicate(
    element: Element,
    private val gameStateToEntity: Function<in GameState, out AbstractEntityState>
) : Predicate<GameState> {

    private val diceRollAdjustmentFormula = element.optionalAttribute("diceRollAdjustment")

    override fun test(gameState: GameState): Boolean {
        val entity = gameStateToEntity.apply(gameState)
        val diceRollResult = rollDice(2)
        val newFormula = diceRollAdjustmentFormula.portOldFormula()
        val finalResult = if (newFormula == null) diceRollResult else keval()
            .withPlayerStateVars(gameState.playerState)
            .withConstant("AMT", diceRollResult.toDouble())
            .eval(newFormula)
            .toInt()
        return finalResult <= entity.skill.currentValue
    }
}